### PR TITLE
use Solr for search

### PR DESCRIPTION
**Pull request by jboning**
This includes no UI. I have a bad UI I've been using to test with, but
it's not ready for prime-time yet. In the interest of keeping changes
small I thought I'd send this for review =)

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -35,7 +35,7 @@
                 'MongoEngine>=0.7.999',
                 'Mako>=0.4.1',
                 'beaker>=1.5',
-                'requests==1.1.0',
+                'requests',
```
Comment (jboning) [inline setup.py:38]: Heads-up here that pysolr wants a recent version of requests, while the yubikey library that core uses wants this old version. If you've been running them out of the same virtualenv, that will no longer work.

File: setup.py [hunk 2]
```diff
@@ -44,7 +44,8 @@
                 'babel',
                 'marrow.mailer',
                 'futures',
-                'bbcode'
+                'bbcode',
+                'pysolr',
```
Comment (jboning) [inline setup.py:48]: pysolr is pretty mediocre. It claims that lxml is an optional requirement, and then if there is a solr error and solr is running on tomcat, it tries to import lxml.html with no fallback, blowing up terribly. So if you want to get good error messages out of solr, it make take some effort.

**Comment by mglinski** [comment]
I know I am a bit late to the party, but if you haven't checked out ElasticSearch you might want to give it a once over. The API is simple JSON data storage and robust enough for a real ranked search service. It provides a binary interface and a full REST interface and has supported clients for all languages.

http://www.elasticsearch.org/guide/en/elasticsearch/guide/current/index.html

The python client does most of the work converting form existing python data types
https://github.com/elasticsearch/elasticsearch-py 

It can also directly pull data and index it from mongo, which might save you guys a bit of time.

File: setup.py (hunk 1 above)
Comment (Acidity) [inline setup.py:38]: D:

**Comment by Acidity** [comment]
Gonna hold on reviewing this until @jboning confirms he doesn't want to use ElasticSearch.

**Comment by jboning** [comment]
ElasticSearch came up as one of the options when I was looking around the internet for comparisons of text search engines. My assessment based on the various writeups I found (mostly on stackoverflow >_>) was that ElasticSearch is easy to distribute over a cluster, but that Solr and Sphinx are easier to use otherwise. The latter seemed more our scale.

I've been pretty happy with what I've seen of Solr so far. Its preference for XML is a little silly, but it speaks JSON if you ask it to. If anyone is strongly in favor of ElasticSearch, I'd be happy to switch. I just didn't see any compelling reason to prefer it.

P.S. you should've chimed in on #53, maybe then I would have gone with ElasticSearch in the first place :P

**Comment by mglinski** [comment]
I honestly don't have a strong preference either way. My only thinking on this is when other apps in the future need robust search functionality, they will most certainly required to us whatever was implemented here. Solr works fine and I have used it before, these engines are almost always built off of Lucene anyway so v0v.

**Comment by jboning** [comment]
Yeah Solr and ElasticSearch are both Lucene (though Sphinx is its own different thing). :shrug: Let's go ahead with this.

**Comment by Acidity** [comment]
Might want to make sure this covers 4d of the Apache License as well

**Comment by jboning** [comment]
There isn't a NOTICE file with the forums right now, so I think we're okay on 4d. (But there probably should be, I'll add an issue to get that sorted.)

**Pull request merged**